File: packages/ios/Sources/Components/CategoryTile/CategoryTile.swift
import SwiftUI

/// A single browsable category cell.
///
/// Mirrors the React `CategoryTile`: the tile owns presentation and selection
/// semantics only. Category identity, labels, counts, and selected state are
/// supplied by the consuming app through `KozmosCategoryPresentation`.
public struct KozmosCategoryTile<Icon: View>: View {
    @Environment(\.kozmosAnalytics) private var trackEvent

    private let category: KozmosCategoryPresentation
    private let isDisabled: Bool
    private let onSelect: (String) -> Void
    private let icon: Icon

    public init(
        category: KozmosCategoryPresentation,
        isDisabled: Bool = false,
        onSelect: @escaping (String) -> Void,
        @ViewBuilder icon: () -> Icon
    ) {
        self.category = category
        self.isDisabled = isDisabled
        self.onSelect = onSelect
        self.icon = icon()
    }

    private var disabled: Bool { category.disabled || isDisabled }

    public var body: some View {
        Button {
            trackEvent(
                KozmosAnalyticsEvent(
                    eventName: "category_selected",
                    component: "CategoryTile",
                    properties: ["categoryId": category.id]
                )
            )
            onSelect(category.id)
        } label: {
            VStack(spacing: KozmosDimensions.primitivesLayoutSpacing75) {
                // The icon's square: 64, radius Control, the container edge;
                // the selection shows on it. The label sits under it.
                icon
                    .frame(width: KozmosDimensions.primitivesLayoutSizing300, height: KozmosDimensions.primitivesLayoutSizing300)
                    .foregroundColor(KozmosColors.primitivesColorsTheme500)
                    .frame(width: KozmosDimensions.primitivesLayoutSizing800, height: KozmosDimensions.primitivesLayoutSizing800)
                    .background(
                        category.selected
                            ? KozmosColors.primitivesColorsTheme500.opacity(0.05)
                            : KozmosColors.primitivesColorsBackground0
                    )
                    .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous))
                    .overlay(
                        RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous)
                            .stroke(
                                category.selected ? KozmosColors.primitivesColorsTheme500 : KozmosColors.semanticsBorderSubtle,
                                lineWidth: category.selected ? 2 : 1
                            )
                    )
                    .accessibilityHidden(true)

                Text(category.label)
                    .font(KozmosTypography.caption2)
                    .foregroundColor(KozmosColors.primitivesColorsForeground100)
                    .multilineTextAlignment(.center)
                    .lineLimit(2)
                    .fixedSize(horizontal: false, vertical: true)

                if let resultCountLabel = category.resultCountLabel {
                    Text(resultCountLabel)
                        .font(KozmosTypography.caption)
                        .foregroundColor(KozmosColors.primitivesColorsForeground500)
                }
            }
            .frame(maxWidth: .infinity)
            .padding(KozmosDimensions.primitivesLayoutSpacing50)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(disabled)
        .opacity(disabled ? 0.5 : 1)
        .accessibilityLabel(category.label)
        .accessibilityValue(category.resultCountLabel ?? "")
        .accessibilityAddTraits(category.selected ? [.isButton, .isSelected] : .isButton)
    }
}

public extension KozmosCategoryTile where Icon == EmptyView {
    init(
        category: KozmosCategoryPresentation,
        isDisabled: Bool = false,
        onSelect: @escaping (String) -> Void
    ) {
        self.init(category: category, isDisabled: isDisabled, onSelect: onSelect) {
            EmptyView()
        }
    }
}

File: packages/android/src/main/java/com/kozmos/components/CategoryTile/CategoryTile.kt
package com.kozmos.components.categorytile

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.draw.clip
import androidx.compose.foundation.border
import androidx.compose.foundation.background
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.material3.LocalContentColor
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.kozmos.components.counter.CounterTone
import com.kozmos.components.counter.KozmosCounter
import com.kozmos.contracts.KozmosCategoryPresentation
import com.kozmos.providers.KozmosAnalyticsEvent
import com.kozmos.providers.LocalKozmosAnalytics
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

/**
 * A single browsable category cell.
 *
 * Mirrors the React `CategoryTile`: the tile owns presentation and selection
 * semantics only. Category identity, labels, counts, and selected state are
 * supplied by the consuming app through [KozmosCategoryPresentation]. A
 * `resultCount` draws as the system's counter at the icon square's top-right;
 * `resultCountLabel` is its spoken form (the state description) and draws
 * nothing. A `tint` — the category's colours — takes the icon, the selection's
 * stroke and the counter's fill and ink; the square stays neutral.
 */
@Composable
fun KozmosCategoryTile(
    category: KozmosCategoryPresentation,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    /** The category's own colour, as the chosen-category field wears it: it
     *  takes the icon and the counter's fill; the square stays neutral. */
    tint: KozmosCategoryTint? = null,
    icon: (@Composable () -> Unit)? = null
) {
    val trackEvent = LocalKozmosAnalytics.current
    val isEnabled = enabled && !category.disabled

    Surface(
        onClick = {
            trackEvent(
                KozmosAnalyticsEvent(
                    component = "CategoryTile",
                    eventName = "category_selected",
                    properties = mapOf("categoryId" to category.id)
                )
            )
            onSelect(category.id)
        },
        modifier = modifier
            .fillMaxWidth()
            .semantics {
                contentDescription = category.label
                selected = category.selected
                category.resultCountLabel?.let { stateDescription = it }
            },
        enabled = isEnabled,
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl),
        color = androidx.compose.ui.graphics.Color.Transparent
    ) {
        Column(
            modifier = Modifier.padding(KozmosDimensions.primitivesLayoutSpacing50),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing75)
        ) {
            // The icon's square: 64, radius Control, the container edge; the
            // selection shows on it. The label sits under it, two lines at most.
            Box {
                Box(
                    modifier = Modifier
                        .size(KozmosDimensions.primitivesLayoutSizing800)
                        .clip(RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
                        .background(
                            if (category.selected) (tint?.accent ?: KozmosColors.primitivesColorsTheme500).copy(alpha = 0.05f)
                            else KozmosColors.primitivesColorsBackground0
                        )
                        .border(
                            width = if (category.selected) 2.dp else 1.dp,
                            color = if (category.selected) (tint?.accent ?: KozmosColors.primitivesColorsTheme500) else KozmosThemeTokens.semanticsBorderSubtle,
                            shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl)
                        ),
                    contentAlignment = Alignment.Center
                ) {
                    if (icon != null) {
                        // The icon in the theme colour, as on the other platforms;
                        // decorative, as there: the label names what it shows.
                        CompositionLocalProvider(LocalContentColor provides (tint?.accent ?: KozmosColors.primitivesColorsTheme500)) {
                            Box(modifier = Modifier.size(KozmosDimensions.primitivesLayoutSizing300).clearAndSetSemantics {}, contentAlignment = Alignment.Center) {
                                icon()
                            }
                        }
                    }
                }
                category.resultCount?.let { count ->
                    // The count: the system's counter, brand tone, at the
                    // square's top-right, four beyond its edges so the icon
                    // stays clear. The spoken form is `resultCountLabel`.
                    KozmosCounter(
                        text = count.toString(),
                        tone = CounterTone.Brand,
                        fill = tint?.fill,
                        modifier = Modifier.align(Alignment.TopEnd).offset(x = 4.dp, y = (-4).dp)
                    )
                }
            }

            Text(
                text = category.label,
                style = MaterialTheme.typography.labelSmall,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
                color = if (isEnabled) {
                    KozmosColors.primitivesColorsForeground100
                } else {
                    KozmosColors.primitivesColorsForeground100.copy(alpha = 0.5f)
                },
                textAlign = TextAlign.Center
            )
        }
    }
}
